File: 2025-12-December-GFG-POTD/December 21 - Count X in Range of a Sorted Array/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> countXInRange(vector<int>& arr, vector<vector<int>>& queries) {
        vector<int> result;

        // Step 1: Preprocess positions
        unordered_map<int, vector<int>> mp;
        for (int i = 0; i < arr.size(); i++) {
            mp[arr[i]].push_back(i);
        }

        // Step 2: Answer queries
        for (auto &q : queries) {
            int l = q[0];
            int r = q[1];
            int x = q[2];

            if (mp.find(x) == mp.end()) {
                result.push_back(0);
                continue;
            }

            vector<int>& indices = mp[x];

            // lower_bound: first >= l
            int left = lower_bound(indices.begin(), indices.end(), l) - indices.begin();

            // upper_bound: first > r
            int right = upper_bound(indices.begin(), indices.end(), r) - indices.begin();

            result.push_back(right - left);
        }

        return result;
    }
};

File: 2025-12-December-GFG-POTD/December 21 - Count X in Range of a Sorted Array/Solution.java
class Solution {
    public ArrayList<Integer> countXInRange(int[] arr, int[][] queries) {
        ArrayList<Integer> result = new ArrayList<>();

        // Step 1: Preprocess positions
        HashMap<Integer, ArrayList<Integer>> map = new HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            map.computeIfAbsent(arr[i], k -> new ArrayList<>()).add(i);
        }

        // Step 2: Answer queries
        for (int[] q : queries) {
            int l = q[0];
            int r = q[1];
            int x = q[2];

            if (!map.containsKey(x)) {
                result.add(0);
                continue;
            }

            ArrayList<Integer> indices = map.get(x);

            int left = lowerBound(indices, l);
            int right = upperBound(indices, r);

            result.add(right - left);
        }

        return result;
    }

    // First index >= target
    private int lowerBound(ArrayList<Integer> list, int target) {
        int low = 0, high = list.size();
        while (low < high) {
            int mid = (low + high) / 2;
            if (list.get(mid) < target)
                low = mid + 1;
            else
                high = mid;
        }
        return low;
    }

    // First index > target
    private int upperBound(ArrayList<Integer> list, int target) {
        int low = 0, high = list.size();
        while (low < high) {
            int mid = (low + high) / 2;
            if (list.get(mid) <= target)
                low = mid + 1;
            else
                high = mid;
        }
        return low;
    }
}
